Dockerlint part 1

Macleykun opened the pull request:
The following programming languages and their solutions have been checked and fixed based on Dockerlint:
PrimeAda
PrimeAmd64
PrimeAssembly
PrimeAssemblyScript
PrimeAWK

Note that somehow, i was unable to build PrimAmd64 and amd64. The dockerfiles look right in my eyes and even with some testing i can see the files. I assume it's just me and i only added versioning on them.

Feel free to review and give feedback, next i'll do another 5 after primeawk :)

rbergen commented on PrimeAssembly/solution_2/Dockerfile:
```diff
@@ -1,15 +1,11 @@
-    ./build.sh
-
-ENTRYPOINT [ "./run.sh" ]
+COPY *.s* ./
```
This is not the same as the COPY commands it seems to replace. *.s* would also copy any other files with an extension starting with .s. There are none right now, but the new copy is less exact.

Macleykun replied:
True! I believed it wouldn't be an issue as it minimized COPY and the less exact doesn't seem an issue to me.
If you wish to prefer exacter COPY instead of less COPY's then i'll do so! It's up to you! I don't have a preference in this :).

rbergen replied:
I deliberately made the COPY as specific as possible when I authored the Dockerfile, to limit the chances of stuff being accidentally pulled into the image at a later stage. So yes, I would prefer the "exact" COPYs over the other one.

Macleykun replied:
Fixed :D

rbergen commented on PrimeAssembly/solution_1/Dockerfile:
```diff
@@ -1,15 +1,11 @@
-COPY *.sh ./
-COPY *.asm ./
-
-RUN chmod a+x ./*.sh && \
-    ./build.sh
```
The chmod a+x is necessary because at the moment, the execute bit is _not_ set on the .sh files in this solution. That's also why the CI run on this PR currently fails.

rbergen commented on PrimeAssembly/solution_2/Dockerfile:
```diff
@@ -1,15 +1,11 @@
-COPY *.sh ./
-COPY *.s ./
-
-RUN chmod a+x ./*.sh && \
-    ./build.sh
```
See previous comment.

rbergen commented:
I'm getting the impression that you're working on WSL2 on Windows, with your local copy of the GitHub repo residing on NTFS. If I'm wrong about that then you can ignore the rest of this comment. If I'm right then you're obviously free to do so, but my recommendation would be to work on an ext4 file system under WSL2, instead.

Using NTFS will lead to all kinds of issues with that filesystem setting executable bits being where they don't belong, and git then removing them when committing, including where they are indeed supposed to be set.

Macleykun commented:
> I'm getting the impression that you're working on WSL2 on Windows, with your local copy of the GitHub repo residing on NTFS. If I'm wrong about that then you can ignore the rest of this comment. If I'm right then you're obviously free to do so, but my recommendation would be to work on an ext4 file system under WSL2, instead.
> 
> Using NTFS will lead to all kinds of issues with that filesystem setting executable bits being where they don't belong, and git then removing them when committing, including where they are indeed supposed to be set.

Well that's something i haven't heard before! I will use a vm instead to do docker testing then! Will fix those .sh files and then those comments are resolved :)

rbergen commented:
> Well that's something i haven't heard before! I will use a vm instead to do docker testing then! Will fix those .sh files and then those comments are resolved :)

I had never heard of it either, until I ran into the exact same thing a few months ago, when getting started on this project. :)

Macleykun commented:
haha i see :) I fixed the one thing ya asked me to do, how does it look?
The second pull will be... quite a bit larger...

rbergen reviewed (approved):
LGTM

rbergen commented:
@Macleykun I'm sorry for not seeing this earlier, but it just occurred to me that you're likely to have used a different hadolint configuration while putting this PR together, than the one we use during our CI. That latter configuration is set in the [config/hadolint.yml](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/config/hadolint.yml) file in the repo.
One issue we ignore in that is 3008, being the one flagging the pinning down of apt-get installs to specific package versions.

I think we'll have to limit the changes in the existing Dockerfiles to those that are reported by hadolint when this project's hadolint configuration is used.

(The reason this now occurs to me is that I'm putting a PR together to add a section concerning hadolint to CONTRIBUTING.md, and I had to figure out how to use custom configuration when running hadolint inside a Docker container.)

Macleykun commented:
Aaah I see, yeah I used default config and fixed everything. It’ll take a bit to re-do all of those but will get to it tonight :).
Would it be possible to merge this maybe soonish after I fixed the issues?

rbergen commented:
For PRs like this, we (the maintainers) consistently apply a double review approach. Because of this, the speed with which PRs can be merged partly and sometimes largely depends on when two of us get round to performing the reviews. I should be able to review an update to the PR pretty quickly but I can't speak on behalf of the others, nor do I want to. It's a matter of nobody getting paid for this, lives to live, day jobs, etc.

Macleykun commented:
true true! tottaly understand! Would it maybe be easier for you two if i segment the pulls even more? I don't mind adjusting to the preference of what's best for the maintainers!

Macleykun commented:
I'll go ahead and go trough all the others i've made, gonna see if can make multiple pulls that are seperate of eachother :P. Then i can atleast keep better up with any changes that are done.

rbergen reviewed (approved):
LGTM, again

rbergen commented:
More PRs aren't necessarily better for us. :) I mean that to the point that if you have other fixes ready of the same types as you apply here, it may actually be best to roll them up in this very PR. I also say that because I really do think we have now identified the right set of changes to make (primarily clean-up after install and consolidating RUNs).

Macleykun commented:
> More PRs aren't necessarily better for us. :) I mean that to the point that if you have other fixes ready of the same types as you apply here, it may actually be best to roll them up in this very PR. I also say that because I really do think we have now identified the right set of changes to make (primarily clean-up after install and consolidating RUNs).

Oh, then i will immediatly start to put all the changes i've made into this PR then! I really though it was a easier, but on second though, it's indeed better to just put it all into one PR. Will have to do another re-review. But hopefully it'll be the last for this PR.

rbergen commented:
Build fails on PrimeCOBOL/solution_1/Dockerfile, due to a change you made to address DL3047. That message is ignored in our config, so that change is not necessary.

Macleykun commented:
> Build fails on PrimeCOBOL/solution_1/Dockerfile, due to a change you made to address DL3047. That message is ignored in our config, so that change is not necessary.

Thanks for pointing that one out! I'm currently going trough all the dockerfiles and testing them, i believed i specfically tested COBOLT but didn't. The issue was an oudated wget which fixed the issue by installing it. I've kinda already addressed DL3047 in the solutions so i'll keep those fixes in.

Every so often i'll push my changes, so technically the CI doesn't have to be re-ran each time. Ofcours it's up to you :).
I'm now at Forth and have about 50 more solutions to go trough :P.

rbergen commented:
I see you've addressed the build failure by bumping wget to a newer version, In line with my earlier comment that changes should in principle be limited to what's needed to address issues flagged by this project's hadolint configuration, I do think the appropriate thing to do is revert the use of the --progress option when running wget, instead.

Macleykun commented:
> I see you've addressed the build failure by bumping wget to a newer version, In line with my earlier comment that changes should in principle be limited to what's needed to address issues flagged by this project's hadolint configuration, I do think the appropriate thing to do is revert the use of the --progress option when running wget, instead.

Addressed in current and upcoming changes!

rbergen commented:
Thanks! I'll hold on my next review and CI run until you notify me that all changes you want to pull into this PR, are indeed in.

Macleykun commented:
Ok this is now ready for CI testing!
Also, i found some left over files i think from vscode?
./PrimeCPP/solution_3/.vscode
./PrimeCPP/solution_2/.vscode

Is it ok if i also delete those in this pull? Or rather not?

rbergen commented on PrimeHaskell/solution_1/Dockerfile:
```diff
@@ -2,19 +2,17 @@ FROM haskell:8.10
 
-COPY src/ /opt/sieve/src/
-COPY app/ /opt/sieve/app/
+COPY src/ app/ ./
```
This change seems to break the build.

Macleykun replied:
Correct, COPY doesn't copy the folder itself, rather it's content. Reverted back to two COPY commands.

rbergen commented on PrimeKos/solution_1/Dockerfile:
```diff
@@ -1,25 +1,17 @@
 
 RUN apk update && \
-    apk add curl make g++
+    apk add --no-cache curl make0 g++
```
This change seems to break the build.

Macleykun replied:
Didn't correctly removed the version tag. Installing make instead of the wrong make0.

rbergen commented on PrimeWren/solution_1/Dockerfile:
```diff
@@ -1,14 +1,14 @@
-    && apt-get install -y curl busybox 
-
-RUN curl -sL https://github.com/wren-lang/wren-cli/releases/download/0.3.0/wren_cli-linux-0.3.0.zip | busybox unzip - -d / \
+    && apt-get install --no-install-recommends -y curl busybox3 \
```
This change seems to break the build.

Macleykun replied:
Same issue as before, didn't removed the version correctly from busybox.
Also it seems that ca-certificates was also needed.

rbergen commented on PrimeProlog/solution_1/Dockerfile:
```diff
@@ -1,13 +1,11 @@
-RUN swipl-ld -cc-options,-w,-O3 -shared -o bitvector bitvector.c
-
-RUN chmod a+x run.sh
+RUN apt-get update && apt-get --no-install-recommends -y install gcc \
```
This change seems to break the build.

Macleykun replied:
libc-dev was also needed.

rbergen commented on PrimeZig/solution_3/Dockerfile:
```diff
@@ -1,9 +1,8 @@
 
 RUN apt-get update && \
-    apt-get install -y curl xz-utils
+    apt-get install --no-install-recommends -y curl xz-utils
```
This change seems to break the build.

Macleykun replied:
ca-certificates had to also be installed

rbergen commented on PrimeV/solution_1/Dockerfile:
```diff
@@ -2,7 +2,7 @@ FROM alpine:3.13 AS build
 ENV V_VER="0.2.4"
 
-RUN apk add --no-cache build-base bash git
+RUN apk update && apk add --no-cache build-base bash git
```
This change seems to break the build (although I have to admit I don't know how.)

Macleykun replied:
Yep, this is something that's sadly unrelated to my pull but.
When i try to compile it with make and gcc on ubuntu. I get the same error.
When i just use the latest, it works.

I'm guessing it has to do with a newer build-base version/gcc/make which doesn't work with version 0.2.4.

You wish to:
use this version: weekly.2021.38 ? It's the only one that works and has a fixed version.
want me (or ask someone else) to find out if an older version of the build tools at 31 august (or earlier) works?

marghidanu replied:
Adding versions to alpine will break more stuff in the future as previous versions are discarded. Also there’s no need to use `apk update` if using `apk add —no-cache`.  Why do we need these improvements for the linter, we can simply ignore more rules.

marghidanu replied:
Also `—no-install-recommends` can skip essential packages, not sure about these rules.

Macleykun replied:
The apk update should be removed. That was unintentional!
--no-install-recommmends avoids installing extra packages which aren't needed. Thus minimizing image size! I've already done this for all the current dockerfiles that i adressed in this pull. I do have to catch up with some newer solutions.

But most issues i've come across is people forgetting to install libc-dev or ca-certificates. It's already done but if you guys don't want to have --no-install-recommends for the future, i'll skip those :-).

marghidanu replied:
I can tell you from experience that `—no-install-recommends` causes problems. I have setup Hadolint for thsi projects :0)

Macleykun replied:
Aah i see, well if you want, i can remove all the occurances of --no-install-recommends if you want, all the builds do work at the moment (exepct the vlang one)

rbergen replied:
I think removing --no-install-recommends is the way to go, particularly if @marghidanu says so. :) Builds working is one thing, but the solutions also need to work when they run. I'm guessing you haven't tested that to exhaustion (if you did, I obviously immediately retract that statement.)

Concerning vlang: if moving up to weekly.2021.38 is the only way to get things working at the current time, then I do think that is the best thing to do.

Macleykun replied:
I though we tested the builds using CI :P 
will make that change for vlang in a second!

rbergen replied:
CI tests builds, not runs.

Macleykun replied:
Removed --no-install-recommends everywhere and also fixed vlang (tested build, it works)

rbergen reviewed (commented):
There are 6 build failures in CI. Did you test these changes locally before committing?

Also, I see some Hadolint findings still remain. Did you mean to leave those?

rbergen commented:
> Is it ok if i also delete those in this pull? Or rather not?

Please don't.

Macleykun commented:
Ok all of the build issues are resolved! The PrimeV one is the only one a decision has to be made on.
I did check every not and then the build, but i figured testing in CI in this pull is quicker/easier. Hence the commited build errors.

Macleykun commented:
The pull is now ready to be reviewed again :-)

rbergen commented:
I think that in some of yesterday's commits. you explicitly added ca-certificates and libc-dev packages to the apt-get installs of some Dockerfiles, to fix build errors when --no-install-recommends was still in. To keep things clean (as in: only change what's needed) these additions should to be reverted.

Macleykun commented:
> I think that in some of yesterday's commits. you explicitly added ca-certificates and libc-dev packages to the apt-get installs of some Dockerfiles, to fix build errors when --no-install-recommends was still in. To keep things clean (as in: only change what's needed) these additions should to be reverted.

Checked all my changes and removed what packages I added, which already get installed with recommended (did this on iPad hence the individual commits).

rbergen commented:
The build of PrimeV/solution_1 fails, again.

Macleykun commented:
> The build of PrimeV/solution_1 fails, again.

I'm not sure how or why this is happening yet, but setting up my vm, the build worked, i pruned everything and rebuilding also fails on my side.

This error is different, i'll try the older version again and will let you know once i have another update.

Macleykun commented:
I've managed to fix it again, by using the latest tag of vlang.
To show i've really, really tested this throughout.
Logs of it working after the fix: https://pastebin.com/w5Ny6Xr2
Logs doing a system prune -a and rebuilding completely again: https://pastebin.com/C51BTgHW

I sadly lack the knowledge/experience with vlang/compilers to really determine why this happens, the only thing that i can imagine is that the dependencies from build-base changed and might cause this behavior.

Anyhow, i'm sorry that it has taken this long to get the builds all working, i prefer finishing this pr and then picking up in a new pr to do the last files so that everything is dockerlint compliant (as configured/desired).

rbergen commented:
I'm comfortable with concluding that you are not at fault here. It seems the relationship between V and its dependencies is extremely fragile, and the build reliability completely horrendous as a consequence. I'm close to concluding we should disable the automated builds of V solutions until the language('s tooling) stabilizes to a point that builds keep working for more than a week.

Macleykun commented:
> I'm comfortable with concluding that you are not at fault here. It seems the relationship between V and its dependencies is extremely fragile, and the build reliability completely horrendous as a consequence. I'm close to concluding we should disable the automated builds of V solutions until the language('s tooling) stabilizes to a point that builds keep working for more than a week.

One proposal i can do is that the dependencies it relies on, can be a fixed version. I know it's not desirable but that might keep the Vlang more stable. Ofcours it's up to you if you prefer to disable the automation for Vlang.

For the future, the dependencies we installed alongside have the versions:
build-base-0.5-r2
git-2.30.2-r0
bash-5.1.0-r0
The full list is: https://pastebin.com/Q4HNbqxz

I'll hear if any changes have to be made :-)

rbergen commented:
Maybe pinning down the versions of dependencies is the way to go in this particular case. If you can apply that to the V Dockerfile then we'll see if we can keep it building until we merge this PR, at least.

Macleykun commented:
> Maybe pinning down the versions of dependencies is the way to go in this particular case. If you can apply that to the V Dockerfile then we'll see if we can keep it building until we merge this PR, at least.

done :-)

rbergen reviewed (approved):
LGTM
